File: 100milliongold/leetcode/tree/100.Same Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
  // 두 노드가 모두 null이면 두 트리의 현재 값이 동일하므로 true
  if (p === null && q === null) {
    return true;
  }
  
  // 한 개의 노드가 null일 때 다른 노드가 null이 아니면 두 노드의 값이 다르므로 false
  if ((p !== null && q === null) || (p === null && q !== null)) {
    return false;
  }
  
  // 두 노드의 값이 서로 다르면 false
  if (p.val != q.val) {
    return false;
  }
  
  // 이진트리를 탐색하므로 왼쪽, 오른쪽을 재귀적으로 순회해서 트리의 왼쪽 자식 노드들과 오른쪽 자식 노드들이 모두 같을 때만 true를 리턴
  return isSameTree(p.right, q.right) && isSameTree(p.left, q.left);
};

File: gunkims/leetcode/tree/100.Same Tree.java
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();

        TreeNode treeNode = new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3, new TreeNode(6), new TreeNode(7)));
        TreeNode treeNode2 = new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3));

        System.out.println(solution.isSameTree(treeNode, treeNode2));
    }
}
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null && q == null) return true;
        if(p == null || q == null) return false;
        if(p.val != q.val) return false;

        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}